File: basic-proxy-impl/src/test/java/com/test/cases/TestBase.java
package com.test.cases;

import com.study.proxy.impl.CodeGenerator;
import com.study.proxy.impl.util.ClassNameProvider;
import com.test.util.MethodLineMatcher;
import com.test.util.PrettyResultBuilder;
import org.junit.Assert;
import org.junit.Test;
import org.objectweb.asm.Type;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Proxy;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.IntStream;

abstract class TestBase {
    private final boolean debug;
    protected final Class<?> specifiedInterface;
    private List<List<String>> standardResult;
    private List<List<String>> realResult;

    /**
     * Filter all lines that start with "//" within a method.
     * Please note that the lines that are before the method name line won't be processed.
     *
     * @param raw raw lines
     * @return the qualified lines
     */
    protected List<String> filter(List<String> raw) {
        int methodStartLine = 0;

        Function<String, Boolean> isCommentedLine = (line) -> line.stripLeading().startsWith("//");

        for (int i = 0; i < raw.size(); i++) {
            String line = raw.get(i);
            if (!isCommentedLine.apply(line)) {
                methodStartLine = i;
                break;
            }
        }

        List<String> result = new ArrayList<>(raw.size());
        result.addAll(raw.subList(0, methodStartLine));
        for (int i = methodStartLine; i < raw.size(); i++) {
            String line = raw.get(i);
            if (!isCommentedLine.apply(line)) {
                result.add(line);
            }
        }
        return result;
    }

    protected void compare(String methodName, String methodDesc) {
        System.out.printf("Comparing method ⬇️%n");
        System.out.printf("name: [%s]%n", methodName);
        System.out.printf("desc: [%s]%n", methodDesc);
        System.out.println();
        List<String> realResult = buildRealResult(methodName, methodDesc);
        List<String> standardResult = buildStandardResult(methodName, methodDesc);

        boolean theSame = theSame(realResult, standardResult);
        Assert.assertTrue(theSame);
    }

    protected TestBase() throws Exception {
        this(false);
    }

    protected TestBase(boolean debug) throws Exception {
        this.debug = debug;
        specifiedInterface = this.getClass().getAnnotation(For.class).value();
        init();
    }

    private void init() throws Exception {
        if (debug) {
            System.out.println("specifiedInterface is: " + specifiedInterface.getName());
        }
        PrettyResultBuilder builder = new PrettyResultBuilder(debug);
        realResult = builder.build(prepareRealBytes(specifiedInterface));
        standardResult = builder.build(prepareStandardBytes(specifiedInterface));
    }

    private byte[] prepareStandardBytes(Class<?> specifiedInterface) throws Exception {
        Proxy.newProxyInstance(
                TestBase.class.getClassLoader(),
                new Class<?>[]{specifiedInterface},
                (proxy, method1, args) -> null
        );
        Method method = Class.forName("java.lang.reflect.ProxyGenerator").
                getDeclaredMethod(
                        "generateProxyClass",
                        ClassLoader.class, String.class, List.class, int.class
                );
        method.setAccessible(true);

        return (byte[]) method.invoke(
                null,
                TestBase.class.getClassLoader(),
                new ClassNameProvider().className(),
                List.of(specifiedInterface),
                Modifier.PUBLIC | Modifier.FINAL
        );
    }

    public byte[] prepareRealBytes(Class<?> specifiedInterface) throws Exception {
        return new CodeGenerator().generate(specifiedInterface);
    }

    private List<String> buildRealResult(String name, String desc) {
        List<String> result = new MethodLineMatcher(realResult).match(name, desc);
        return filter(result);
    }

    private List<String> buildStandardResult(String name, String desc) {
        List<String> result = new MethodLineMatcher(standardResult).match(name, desc);
        return filter(result);
    }

    private boolean theSame(List<String> realResult, List<String> standardResult) {
        int size1 = realResult.size();
        int size2 = standardResult.size();
        if (size1 != size2) {
            System.out.printf("size1 is: %s, size2 is: %s%n", size1, size2);
//            return false;
        }
        int size = Math.min(size1, size2);
        for (int i = 0; i < size; i++) {
            String line1 = realResult.get(i);
            String line2 = standardResult.get(i);
            if (!Objects.equals(line1, line2)) {
                System.out.printf("Difference detected in line: [%s]%n", i);
                System.out.println("Real result:");
                System.out.println(line1);
                System.out.println("Standard result:");
                System.out.println(line2);
                return false;
            }
            if (debug) {
                System.out.println(line1);
            }
        }
        boolean theSame = size1 == size2;
//        if (debug) {
        System.out.printf("size is: %s, is the content the same?: [%s]%n", size, theSame);
//        }
        return theSame;
    }

    @Test
    public void testBasicMethod() {
        List<Map.Entry<String, String>> list = List.of(
                Map.entry("<init>", "(Ljava/lang/reflect/InvocationHandler;)V"),
                Map.entry("hashCode", "()I"),
                Map.entry("equals", "(Ljava/lang/Object;)Z"),
                Map.entry("toString", "()Ljava/lang/String;"),
                Map.entry("<clinit>", "()V")
        );
        System.out.println("===> [Basic Methods Comparison] Starts");
        IntStream.range(0, list.size()).forEach(index -> {
            Map.Entry<String, String> entry = list.get(index);
            String methodName = entry.getKey();
            String methodDesc = entry.getValue();
            compare(methodName, methodDesc);
        });
        System.out.println("<=== [Basic Methods Comparison] Ends");
    }

    @Test
    public void testAllDeclaredMethods() {
        Function<Method, Boolean> isBasicMethod = (method -> {
            if (method.getName().equals("hashCode") && Type.getMethodDescriptor(method).equals("()I")) {
                return true;
            }
            if (method.getName().equals("equals") && Type.getMethodDescriptor(method).equals("(Ljava/lang/Object;)Z")) {
                return true;
            }
            return method.getName().equals("toString") && Type.getMethodDescriptor(method).equals("()Ljava/lang/String;");
        });

        System.out.println("===> [Declared Methods Comparison] Starts");
        for (Method method : specifiedInterface.getDeclaredMethods()) {
            if (method.isSynthetic()) {
                continue;
            }
            if (Modifier.isStatic(method.getModifiers())) {
                continue;
            }
            if (method.isDefault()) {
                continue;
            }
            if (isBasicMethod.apply(method)) {
                continue;
            }

            compare(method.getName(), Type.getMethodDescriptor(method));
        }
        System.out.println("<=== [Declared Methods Comparison] Ends");
    }
}

File: basic-proxy-kotlin-impl/src/test/kotlin/com/test/cases/TestBase.kt
package com.test.cases

import com.study.proxy.impl.CodeGenerator
import com.study.proxy.impl.util.ClassNameProvider
import com.test.util.MethodLineMatcher
import com.test.util.PrettyResultBuilder
import org.junit.Assert
import org.junit.Test
import org.objectweb.asm.Type
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.lang.reflect.Proxy
import kotlin.reflect.KClass

abstract class TestBase(val debug: Boolean) {
    protected val specifiedInterface: KClass<*> = this::class.java.getAnnotation(For::class.java).value
    private var standardResult: List<List<String>>
    private var realResult: List<List<String>>

    constructor() : this(false)

    init {
        if (debug) {
            println("specifiedInterface is: ${specifiedInterface.java.getName()}")
        }
        val builder = PrettyResultBuilder(debug)
        realResult = builder.build(prepareRealBytes(specifiedInterface.java))
        standardResult = builder.build(prepareStandardBytes(specifiedInterface.java))
    }

    /**
     * Filter all lines that start with "//" within a method.
     * Please note that the lines that are before the method name line won't be processed.
     *
     * @param raw raw lines
     * @return the qualified lines
     */
    protected fun filter(raw: List<String>): List<String> {
        var methodStartLine = 0

        val isCommentedLine = { line: String -> line.trimStart().startsWith("//") }

        for (i in 0 until raw.size) {
            val line = raw[i]
            if (!isCommentedLine(line)) {
                methodStartLine = i
                break
            }
        }

        val result = mutableListOf<String>()
        result.addAll(raw.subList(0, methodStartLine))
        raw.subList(methodStartLine, raw.size).filter { !isCommentedLine(it) }.forEach { result.add(it) }
        return result
    }

    protected fun compare(methodName: String, methodDesc: String) {
        println("Comparing method ⬇️")
        println("name: [${methodName}]")
        println("desc: [${methodDesc}]")
        println()
        val realResult = buildRealResult(methodName, methodDesc)
        val standardResult = buildStandardResult(methodName, methodDesc)

        val theSame = theSame(realResult, standardResult)
        Assert.assertTrue(theSame)
    }

    private fun prepareStandardBytes(specifiedInterface: Class<*>): ByteArray {
        Proxy.newProxyInstance(
            TestBase::class.java.getClassLoader(),
            arrayOf(specifiedInterface)
        ) { proxy, method, args -> TODO("Not yet implemented") }
        val method = Class.forName("java.lang.reflect.ProxyGenerator").getDeclaredMethod(
            "generateProxyClass",
            ClassLoader::class.java, String::class.java, List::class.java, Integer.TYPE
        )
        method.setAccessible(true)

        return method.invoke(
            null,
            TestBase::class.java.getClassLoader(),
            ClassNameProvider.className(),
            listOf(specifiedInterface),
            Modifier.PUBLIC or Modifier.FINAL
        ) as ByteArray
    }

    fun prepareRealBytes(specifiedInterface: Class<*>): ByteArray {
        return CodeGenerator().generate(specifiedInterface)
    }

    private fun buildRealResult(name: String, desc: String): List<String> {
        val result = MethodLineMatcher(realResult).match(name, desc)
        return filter(result)
    }

    private fun buildStandardResult(name: String, desc: String): List<String> {
        val result = MethodLineMatcher(standardResult).match(name, desc)
        return filter(result)
    }

    private fun theSame(realResult: List<String>, standardResult: List<String>): Boolean {
        val size1 = realResult.size
        val size2 = standardResult.size
        if (size1 != size2) {
            println("size1 is: ${size1}, size2 is: $size2")
//            return false
        }
        val size = size1.coerceAtMost(size2)
        for (i in 0 until size) {
            val line1 = realResult[i]
            val line2 = standardResult[i]
            if (!java.util.Objects.equals(line1, line2)) {
                println("Difference detected in line: [${i}]")
                println("Real result:")
                println(line1)
                println("Standard result:")
                println(line2)
                return false
            }
            if (debug) {
                println(line1)
            }
        }
        val theSame = size1 == size2
//        if (debug) {
        println("size is: ${size}, is the content the same?: [${theSame}]")
//        }
        return theSame
    }

    @Test
    fun testBasicMethod() {
        val map: Map<String, String> = mapOf(
            "<init>" to "(Ljava/lang/reflect/InvocationHandler;)V",
            "hashCode" to "()I",
            "equals" to "(Ljava/lang/Object;)Z",
            "toString" to "()Ljava/lang/String;",
            "<clinit>" to "()V"
        )
        println("===> [Basic Methods Comparison] Starts")
        map.forEach {
            val methodName = it.key
            val methodDesc = it.value
            compare(methodName, methodDesc)
        }
        println("<=== [Basic Methods Comparison] Ends")
    }

    @Test
    fun testAllDeclaredMethods() {
        val isBasicMethod = { method: Method ->
            val name = method.name
            val descriptor = Type.getMethodDescriptor(method)

            val isHashCodeMethod = name.equals("hashCode") && descriptor.equals("()I")
            val isEqualsMethod = name.equals("equals") && descriptor.equals("(Ljava/lang/Object;)Z")
            val isToStringMethod = name.equals("toString") && descriptor.equals("()Ljava/lang/String;")

            isHashCodeMethod || isEqualsMethod || isToStringMethod
        }

        println("===> [Declared Methods Comparison] Starts")
        for (method in (this as Any).javaClass.getDeclaredMethods()) {
            if (method.isSynthetic) {
                continue
            }
            if (Modifier.isStatic(method.modifiers)) {
                continue
            }
            if (method.isDefault) {
                continue
            }
            if (isBasicMethod(method)) {
                continue
            }

            compare(method.name, Type.getMethodDescriptor(method))
        }
        println("<=== [Declared Methods Comparison] Ends")
    }
}
